### PR TITLE
Increase the output color of search results

### PR DESCRIPTION
**Pull request by QiangHeisenberg**
close https://github.com/rust-lang/cargo/issues/9918

### CODE REVIEW

**Comment by QiangHeisenberg** [comment]
Why the test task has not passed? I'm still looking for the reason. If anyone knows, feel free to let me know.

**Comment by ehuss** [comment]
> Why the test task has not passed?

`status()` goes to stderr, it can't be used for displaying the search results to stdout.

**Comment by QiangHeisenberg** [comment]
> > Why the test task has not passed?
> 
> `status()` goes to stderr, it can't be used for displaying the search results to stdout.

thanks. I didn't notice, I was too careless. The failed test now seems to have nothing to do with me or something.

**Comment by ehuss** [comment]
Sorry if I wasn't clear. The tests can't change to check stderr, the code will need to be changed to use stdout. I don't think the `status` function can be used here.

I opened #10120 to track the spurious CI error.

**Comment by QiangHeisenberg** [comment]
Thanks, you are right.
The two are indeed different, and the test code cannot be changed rashly.

Currently there is an output method for stderr, but there is no corresponding method for stdout. As a supplement, I added a corresponding output method for stdout.

**Comment by ehuss** [comment]
We talked about this a bit, and felt like coloring the entire entry in bright green can make it a bit difficult to read and doesn't particularly improve the results. We were thinking maybe it could just highlight the matching terms in a normal green color in a fashion similar to `npm search`.  For example `cargo search foo` would highlight the term `foo` wherever it appears in the line.  

Would you be willing to try that out?  It will require a bit more work to figure out a good API for the `Shell`.  These aren't "status" messages precisely. I don't know what that API would look like. I think it would be good to be somewhat flexible (being able to pass a string with specific parts specified with color), but I don't want to make it too complex. For example, there are crates like `colored` which make it easier, but I would prefer to not bring in a crate just for this.  If there is a good balance of simple and flexible, that would be good.

**Comment by QiangHeisenberg** [comment]
thank you. I have made the latest changes. I haven't figured out how to design test cases. However, I don't see any other commands to test the output color. It seems unnecessary.

**Comment by alexcrichton** [comment]
Can you post a screenshot or two of what the search results look like?

**Comment by QiangHeisenberg** [comment]
thank you, of course.
image

image

image

File: src/cargo/core/shell.rs [hunk 1]
```diff
@@ -143,6 +143,27 @@ impl Shell {
...
+    /// Print the message to standard output, where the query will have the color `color`.
+    pub fn print_stdout_with_part_color(
+        &mut self,
+        name_no_query: Vec<&str>,
+        desc_no_query: Vec<&str>,
+        query: &str,
+        color: Color,
+        justified: bool,
+    ) -> CargoResult<()> {
+        match self.verbosity {
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs:155]: Wouldn't this mean that `cargo search -q foo` wouldn't actually print anything?

File: src/cargo/core/shell.rs [hunk 2]
```diff
@@ -423,6 +454,63 @@ impl ShellOut {
...
     }
 
+    /// Prints out a message with a status to stdout. Output the specified color and bold for the query content.
+    fn message_stdout(
+        &mut self,
+        name_no_query: Vec<&str>,
+        desc_no_query: Vec<&str>,
+        query: &str,
+        color: Color,
+        justified: bool,
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs:464]: This is always `true` right now so does this need to be a parameter?

File: src/cargo/core/shell.rs [hunk 3]
```diff
@@ -204,6 +225,16 @@ impl Shell {
         self.print(&status, Some(&message), Green, true)
     }
 
+    /// Shortcut to right-align and paint the query part of the output message as a green in stdout.
+    pub fn status_stdout_part_green(
+        &mut self,
+        name_no_query: Vec<&str>,
+        desc_no_query: Vec<&str>,
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs:232]: Can you document what these arguments are? I can't figure out what these are from the signature/name of the method...

File: src/cargo/core/shell.rs [hunk 4]
```diff
@@ -204,6 +225,16 @@ impl Shell {
         self.print(&status, Some(&message), Green, true)
     }
 
+    /// Shortcut to right-align and paint the query part of the output message as a green in stdout.
+    pub fn status_stdout_part_green(
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs:229]: Given how specific this method is to `cargo search` I think it would be better to implement this in the `cargo search` area rather than in the general shell area.

File: src/cargo/core/shell.rs [hunk 5]
```diff
@@ -423,6 +454,63 @@ impl ShellOut {
...
+                    write!(stdout, "{}", message)?;
+                    stdout.set_color(ColorSpec::new().set_bold(true).set_fg(Some(color)))?;
+                    if count_name != name_no_query.len() {
+                        write!(stdout, "{}", query)?;
+                    }
+                }
+
+                if !desc_no_query.is_empty() {
+                    let mut count_desc = 0;
+                    for message in desc_no_query.iter() {
```
Comment (alexcrichton) [inline src/cargo/core/shell.rs:481]: This looks to be the same loop as above, so can the code be refactored to avoid this duplication?

File: src/cargo/core/shell.rs (hunk 1 above)
Comment (QiangHeisenberg) [reply src/cargo/core/shell.rs]: thank you,done.

File: src/cargo/core/shell.rs (hunk 3 above)
Comment (QiangHeisenberg) [reply src/cargo/core/shell.rs:224]: Added explanation and changed parameter naming

File: src/cargo/core/shell.rs (hunk 4 above)
Comment (QiangHeisenberg) [reply src/cargo/core/shell.rs:221]: thank you, i have changed it.

File: src/cargo/core/shell.rs (hunk 5 above)
Comment (QiangHeisenberg) [reply src/cargo/core/shell.rs]: thank you，refactored the code.

File: src/cargo/core/shell.rs (hunk 2 above)
Comment (QiangHeisenberg) [reply src/cargo/core/shell.rs:458]: Because there are cases where color output is not supported. 
image

I learned the way of the ```message_stderr()```function.

**Comment by QiangHeisenberg** [comment]
@ehuss @alexcrichton @Eh2406 Do you mind having a look when you get a chance? I want to know if I am on the right track. Thanks!

**Comment by alexcrichton** [comment]
Sorry I do not personally have time to shepherd along this PR.

**Comment by ehuss** [comment]
@QiangHeisenberg Sorry for the lack of direction, but I don't think this is quite the way I was thinking.  The way this is written seems to be highly specific to the needs for the search results.  I was expecting something that was a little more generic that could be used by other things in the future.

For example, I think the `Shell` method could be something like `write_stdout(text: &str, color: ColorSpec) -> CargoResult<()>`.  Then the search code could split the string on the matched text, and alternate between normal and green colors.  There shouldn't be anything related to justification or status messages in this low-level API.

I also want to say that these reviews are taking an outsized amount of time from us. I would appreciate a little more effort on your side to understand how to approach these problems, and to respond to review comments. Alex asked why `message_stdout` takes a `justified` parameter if it is always true.  The response to that was not clear, and I also don't see why it took that parameter, as I also see it always taking a `true` value.  I also find the API in this PR a little hard to follow, and the method parameter names are not enough to make that clear.

Finally, I would expect some kind of test to validate the behavior with color.  That can be done by using `--color=always`, and there are a few examples in the testsuite doing that.

**Comment by epage** [comment]
I've gone ahead and created https://github.com/rust-lang/cargo/pull/10425 that tries to resolve @ehuss's concerns as part of an experiment for more flexible colored output for cargo-add.  It works alright in this case.

**Comment by alexcrichton** [comment]
I've r+'d https://github.com/rust-lang/cargo/pull/10425, so I'm going to close this.
